File: backend/app/schemas/analytics.py
from pydantic import BaseModel


class TrackEventRequest(BaseModel):
    event_type: str
    token: str | None = None
    user_id: int | None = None
    metadata: dict | None = None


class AnalyticsSummary(BaseModel):
    event_type: str
    count: int

File: frontend/src/lib/analytics.ts
import { api, apiPublic } from "./api";
import { getToken } from "./auth";

export async function trackEvent(
  eventType: string,
  opts?: { token?: string; userId?: number; metadata?: Record<string, unknown> }
): Promise<void> {
  const body = {
    event_type: eventType,
    token: opts?.token,
    user_id: opts?.userId,
    metadata: opts?.metadata,
  };
  try {
    if (getToken()) {
      await api.post("/api/analytics/track-event", body);
    } else {
      await apiPublic.post("/api/analytics/track-event/public", body);
    }
  } catch {
    /* non-blocking */
  }
}
